indispensable/src/commonMain/kotlin/at/asitplus/signum/indispensable/asn1/ObjectIdentifier.kt

```diff
@@ -19,48 +24,125 @@ import kotlinx.serialization.encoding.Encoder
  * @throws Asn1Exception if less than two nodes are supplied, the first node is >2 or the second node is >39
  */
 @Serializable(with = ObjectIdSerializer::class)
-class ObjectIdentifier @Throws(Asn1Exception::class) constructor(@Transient vararg val nodes: UInt) :
+class ObjectIdentifier @Throws(Asn1Exception::class) private constructor(
+    bytes: ByteArray?,
+    nodes: List<BigInteger>?
+) :
     Asn1Encodable<Asn1Primitive> {
-
     init {
-        if (nodes.size < 2) throw Asn1StructuralException("at least two nodes required!")
-        if (nodes[0] * 40u > UByte.MAX_VALUE.toUInt()) throw Asn1Exception("first node too lage!")
-        //TODO more sanity checks
+        if ((bytes == null) && (nodes == null)) {
+            //we're not even declaring this, since this is an implementation error on our end
+            throw IllegalArgumentException("either nodes or bytes required")
+        }
+        if (bytes?.isEmpty() == true || nodes?.isEmpty() == true)
+            throw Asn1Exception("Empty OIDs are not supported")
+
+        bytes?.apply {
+            if(first().toUByte()>127u) throw Asn1Exception("OID top-level arc can only be number 0, 1 or 2")
+        }
+        nodes?.apply {
+            if (size < 2) throw Asn1StructuralException("at least two nodes required!")
```

Comment:
does this check not apply to the bytearray version?

Reply:
the first two nodes are encoded into a single byte (the first one), so you need discrete checks, dependeing on where you are coming from